### PR TITLE
Make both ALM and EPM state type stable.

### LINKED ISSUE #242: Stabilize types in `AugmentedLagrangianMethodState`
Currently `AugmentedLagrangianMethodState` has almost all of its fields abstractly typed. This can hurt performance. Can we get most (or all of them) concretely typed? I can make a PR that adds a few new type parameters for necessary types.

### PR DESCRIPTION
**Pull request by kellertuer**
This fixes #242 by introducing a type for the real fields and for ALM one for the two vectors (of same field type basically).

### CODE REVIEW

File: src/solvers/augmented_Lagrangian_method.jl [hunk 1]
```diff
@@ -40,47 +40,54 @@ in the keyword arguments.
 [`augmented_Lagrangian_method`](@ref)
 """
 mutable struct AugmentedLagrangianMethodState{
-    P,Pr<:AbstractManoptProblem,Op<:AbstractManoptSolverState,TStopping<:StoppingCriterion
+    P,
+    Pr<:AbstractManoptProblem,
+    Op<:AbstractManoptSolverState,
+    R,
```
Comment (mateuszbaran) [inline src/solvers/augmented_Lagrangian_method.jl:46]:
```suggestion
    R<:Real,
```

File: src/solvers/augmented_Lagrangian_method.jl [hunk 2]
```diff
@@ -40,47 +40,54 @@ in the keyword arguments.
...
         ϵ_exponent=1 / 100,
         θ_ϵ=(ϵ_min / ϵ)^(ϵ_exponent),
-        stopping_criterion::StoppingCriterion=StopAfterIteration(300) | (
+        stopping_criterion::SC=StopAfterIteration(300) | (
             StopWhenSmallerOrEqual(:ϵ, ϵ_min) & StopWhenChangeLess(1e-10)
         ),
-    ) where {P,Pr<:AbstractManoptProblem,St<:AbstractManoptSolverState}
-        alms = new{P,Pr,St,typeof(stopping_criterion)}()
+    ) where {
+        P,Pr<:AbstractManoptProblem,R,V,SC<:StoppingCriterion,St<:AbstractManoptSolverState
```
Comment (mateuszbaran) [inline src/solvers/augmented_Lagrangian_method.jl:88]:
```suggestion
        P,Pr<:AbstractManoptProblem,R<:Real,V,SC<:StoppingCriterion,St<:AbstractManoptSolverState
```

File: src/solvers/exact_penalty_method.jl [hunk 3]
```diff
@@ -30,41 +30,43 @@ arguments.
 
 [`exact_penalty_method`](@ref)
 """
-mutable struct ExactPenaltyMethodState{P,Pr,Op,TStopping<:StoppingCriterion} <:
+mutable struct ExactPenaltyMethodState{P,Pr,St,R,TStopping<:StoppingCriterion} <:
```
Comment (mateuszbaran) [inline src/solvers/exact_penalty_method.jl:33]:
```suggestion
mutable struct ExactPenaltyMethodState{P,Pr,St,R<:Real,TStopping<:StoppingCriterion} <:
```

File: src/solvers/exact_penalty_method.jl [hunk 4]
```diff
@@ -30,41 +30,43 @@ arguments.
...
+        ρ::R=1.0,
+        θ_ρ::R=0.3,
+        stopping_criterion::SC=StopWhenAny(
             StopAfterIteration(300),
             StopWhenAll(StopWhenSmallerOrEqual(:ϵ, ϵ_min), StopWhenChangeLess(1e-10)),
         ),
-    ) where {P,Pr<:AbstractManoptProblem,Op<:AbstractManoptSolverState}
-        epms = new{P,Pr,Op,typeof(stopping_criterion)}()
+    ) where {
+        P,Pr<:AbstractManoptProblem,St<:AbstractManoptSolverState,R,SC<:StoppingCriterion
```
Comment (mateuszbaran) [inline src/solvers/exact_penalty_method.jl:67]:
```suggestion
        P,Pr<:AbstractManoptProblem,St<:AbstractManoptSolverState,R<:Real,SC<:StoppingCriterion
```

**Review by mateuszbaran — APPROVED** [review]
Just a few type bounds that look reasonable.

**Comment by kellertuer** [comment]
Good idea, I left that open initially due to complex manifolds, but you are right, for penalties, that should all be reals.

**Comment by kellertuer** [comment]
Wah. This PR got a bit stronge, it seems we managed to get two Format dogs hunting each other. Let's wait a bit.

**Pull request merged**